cli/main.c

```diff
@@ -1377,8 +1394,9 @@ static int fw_update(int argc, char **argv)
 	}
 
 	progress_start();
-	ret = switchtec_fw_write_file(cfg.dev, cfg.fimg, cfg.dont_activate,
-				      cfg.force, progress_update);
+	ret = switchtec_fw_write_file_ex(cfg.dev, rpc_cmd,
```

Comment:
I don't like the structure here... I don't think any users trying to update firmware should have to deal with raw MRPC commands or boot phases.

Instead of creating a bunch of new public functions, with the _ex prefix, create them as static (and instead of naming them with an _ex prefix, it's common to just prepend two underscores to the name).

Then have switchtec_fw_write_file() (and others) check the boot phase before performing the underlying operation. Then you don't have to make fw_update() or other callers, any more complicated; and our overall API isn't made more complicated either.

Also, if we are checking the boot phase often, it *may* make sense just to load this information and cache it on switchtec_open()...

Reply:
There are several constraints at play here:
- existing API should continue to be supported
- new BL2 fw-update feature only differs by command ID, we want minimal duplicate code
- function _switchtec_fw_wait()_ and _switchtec_fw_dlstatus()_ are called from _for_ loops, for performance considerations, we want to avoid checking boot phase in these functions
- from API perspective, it is possible (though less likely) that user open device at one boot phase, and do fw-update in another boot phase, so the cached boot phase becomes stale

I agree that users should not have to deal with MRPC commands and boot phases. We can move the boot phase checking and MRPC command to _switchtec_fw_write_file()_. But the underlying functions _switchtec_fw_wait()_ and _switchtec_fw_dlstatus()_ has to be extended to accept MRPC command ID.

Reply:
Updated to reflect my suggestion above. The function name _switchtec_fw_wait_ex()_ and _switchtec_fw_dlstatus_ex()_ can be changed as appropriate. 

Please take a look again, thanks.

Reply:
I don't really see any issue here.

Probably one mistake is that switchtec_fw_wait() and switchtec_fw_dlstatus() shouldn't have ever been exported. They aren't used by anything. 

Probably the easiest solution is to cache the boot phase in the device structure. 

Another solution is to just have the public switchtec_fw_wait() and switchtec_fw_dlstatus() call the private version which take the mrpc command. Though in this case, using them will probably be pretty broken.

Just avoid introducing a bunch of new complicated public APIs for this...

Reply:
I have updated code based on the 2nd solution above, considering that a phase change is still possible between _switchtec_open()_ and fw-update (to confirm by Kelvin). 

No new API function is introduced. The public switchtec_fw_wait() and switchtec_fw_dlstatus() just calls the private version with known MRPC command. 

If this update is OK, I'll wait for comments from Kelvin.

Reply:
@lsgunth We will have another mfg command/API boot-resume which makes the firwmare step into another boot phase (BL1->BL2->Main). As different boot phase runs different F/W image with different command set, so the run time environment across this API will be different. 

I think we have two options, either forcing the fd be invalid after a boot-resume (so user has to do close and re-open after the boot phase switch), and caching the boot phase on open; or get the boot phase locally to any function that may branch for different phase. I guess option 1 is probably better as other stuff like device node name and pax id may also change over a resume, continue using the old fd may cause unknown side effect.

Reply:
My vote goes to option 1 as well. If we can make sure user re-open the dev after switching boot phase, then our code can be much cleaner.

Reply:
I think the complaint is now the unused switchtec_fw_wait() and switchtec_fw_dlstatus() are pretty broken. We could probably safely drop them though.

I think operations like a boot phase switch will need to have an exclusive lock on the device... Sounds like it will be problematic if anything else is using the device while it occurs. So on switchtec_open() we will need to get a shared flock, then on exclusive operations like a boot phase change we'd switch to an exclusive flock which ensures no other processes are using it. Thus caching the boot phase is now an option because it can't change while you have the device open unless your own process changed it.

I'm actually a bit surprised the firmware can change phase without requiring a host to be rebooted. It sounds like a pretty drastic thing.

Reply:
Code updated based on Kelvin's option 2. @lsgunth please take a look.

Reply:
Yes, this looks much cleaner, but I'm still against creating a second firmware update API and CLI command.

Reply:
I agree we can drop the switchtec_fw_dlstatus() and switchtec_fw_wait() from the API list, with a major version update. I don't see any chance user should use it.